web/src/desktop_notifications.ts

```diff
@@ -0,0 +1,107 @@
+import $ from "jquery";
+
+import type {ElectronBridge} from "./electron_bridge";
+import type {Message} from "./message_store";
+
+const electron_bridge: ElectronBridge | undefined = window.electron_bridge;
+class ElectronBridgeNotification extends EventTarget {
+    constructor(title: string, options: NotificationOptions) {
+        super();
+        if (
+            electron_bridge &&
+            electron_bridge.new_notification &&
+            typeof electron_bridge.new_notification === "function"
+        ) {
+            Object.assign(
+                this,
+                electron_bridge.new_notification(
+                    title,
+                    options,
+                    (type: string, eventInit: EventInit) =>
+                        this.dispatchEvent(new Event(type, eventInit)),
+                ),
+            );
+        }
+    }
+    static get permission(): NotificationPermission {
+        return Notification.permission;
+    }
+    static async requestPermission(
+        callback: NotificationPermissionCallback | undefined = undefined,
+    ): Promise<NotificationPermission> {
+        if (callback) {
+            callback(await Promise.resolve(Notification.permission));
+        }
+        return Notification.permission;
+    }
+    close(): void {
+        this.close();
+    }
```

Comment:
This is an infinite recursive loop!!

Reply:
**Reason of creating this function** : 
When the new instance of the class is created, ```close``` function is inside the object created by ```Object.assign()``` inside the constructor. That ```close()``` function can't be called directly like ```instance_of_class.close()``` in the TypeScript (In JavaScript, it can be called directly in that way but not in TypeScript). If we try to call it like that way, it will throw an error that "Property ```close``` doesn't exist on this type". That's why I created this function. 

**How this function will work without causing infinite loop** : 
When ```instance_of_class.close()``` is called, it will call ```this.close()``` which is the ```close()``` function inside the object created by constructor. The ```close()``` function always exist in the object. So it will not lead to an infinite loop.

**When It can be lead to infinite loop**:
If the close function is not exist inside the object created by constructor, it will lead to an infinite loop. If we remove this function to avoid that, there will be error in several places that 'Property ```close``` doesn't exist'.

**How I will modify the code to avoid both errors**:
I will add an ```if``` statement before calling ```this.close()``` to check that "close" function exist inside the object created by constructor or not. It will only call ```this.close()``` if the function exists, this will make sure that function not goes into infinite loop even if the 'close' function is not in the object and also make sure that code doesn't throw 'error : close property doesn't exist'.

Reply:
I have changed the code now and added an ```if``` statement that check specifically in the object that the ```close``` function exists or not. If the function exists in the object then that function will be called and there will be no infinite loop.